WP REST API: Add VideoPress data to media endpoints

mmtr opened the pull request:
#### Changes proposed in this Pull Request:

This PR adds a new plugin that extends the media endpoints (`/wp/v2/media`) including VideoPress data so it can be used by Gutenberg (which retrieves data using `apiFetch` from `/wp/v2/media`). This plugin will be in sync with WordPress.com (D23321-code).

This is required by the VideoPress extension we're building for Gutenberg, since the VideoPress data is available currently only on `/rest/v1.1` endpoints that don't work with Jetpack sites.

#### Questions

I'm worried about the performance cost of this change. Do we have any way of checking it? Should I create an independent endpoint for retrieving the VideoPress data?

#### Testing instructions:
* Apply D23321-code, sandbox the API and create a new Jetpack site using JN: https://jurassic.ninja/create?jetpack-beta&branch=add/rest-api-media-videopress.
* Connect the Jetpack site to WordPress.com using a premium/business subscription.
* Go to Media > Library and upload an image and a video.
* Go to Jetpack > Settings and activate the VideoPress module (by enabling the "Enable high-speed, ad-free video player" option).
* Go to Media > Library and upload a new video.
* Perform a request to the media endpoint: `GET https://public-api.wordpress.com/wp/v2/sites/:site/media`.
* Check that the `jetpack_videopress_guid` property is included only on the video items (it should be an empty string for the video uploaded before activating VideoPress and the GUID value for the one uploaded after activating it).
* Deactivate the VideoPress module or remove the premium/business subscription.
* Perform again the request to the media endpoint.
* Make sure the `jetpack_videopress_guid` is still present on the video items.

#### Proposed changelog entry for your changes:

* WP REST API: Add VideoPress data to media endpoints

mmtr committed:
WP REST API: Add VideoPress data to media endpoints

kwight commented on _inc/lib/core-api/wpcom-endpoints/sites-media-videopress.php:
```diff
@@ -0,0 +1,33 @@
+		return $videopress_data;
+	}
+}
+wpcom_rest_api_v2_load_plugin( 'WPCOM_REST_API_V2_Sites_Media_VideoPress' );
```
Can we check that VideoPress is active before calling this? `./subscribers.php` has an example.

mmtr replied:
Done in 3229832, but since the goal is to have this plugin synced with WP.com, I included additional checks making sure we're in the right environment. I applied the same changes in D23321-code.

kwight commented:
Should this be in `wpcom-fields` instead of `wpcom-endpoints`? That seems more appropriate, I think (although I couldn't really find any documentation on the difference, this just seems to be doing what `_inc/lib/core-api/wpcom-fields/post-fields-publicize-connections.php` is doing too).

jeherve commented:
> Should this be in wpcom-fields instead of wpcom-endpoints? That seems more appropriate, I think (although I couldn't really find any documentation on the difference

@mdawaffe should be able to shed some light on this.

mmtr committed:
Check if VideoPress module is active

mmtr commented:
> Should this be in `wpcom-fields` instead of `wpcom-endpoints`? That seems more appropriate, I think (although I couldn't really find any documentation on the difference, this just seems to be doing what `_inc/lib/core-api/wpcom-fields/post-fields-publicize-connections.php` is doing too).

I don't think so. We already have `sites-posts-featured-media-url.php` in `wpcom-endpoints` which also includes a new field on the API REST responses.

mmtr commented:
> Taking a quick look at the process and returns of various VP functions, you could simplify `get_videopress_data` to probably just:
> 
> ```
> $videopress_id = get_post_meta( $object['id'], 'videopress_guid', true );
> return videopress_get_video_details( $videopress_id );
> ```

I'm afraid this is not possible, because this plugin will be synced with WP.com and we store the VideoPress data differently there (in a different table rather than as post metadata). 

`videopress_get_video_details` and `videopress_get_video_details` are the only functions we can safely use in both environments.

kwight commented:
Awesome, I checked this on both a JN site, and dotcom, everything worked according to the instructions.

I did notice that on the JN site, upgrading the plan resulted in a new field, `author_wpcom`, for existing media items. New VideoPress media items, however, don't have this field; I'm not sure if that's related, or a separate bug (or not?) altogether. Also, two media items were created by just one upload, one with JP data and one without – this doesn't happen on dotcom, however. Oh! One is a jpeg, so maybe a poster image of some sort?

<img width="1186" alt="screen shot 2019-01-23 at 9 02 09 am" src="https://user-images.githubusercontent.com/349751/51624056-ee441600-1eee-11e9-9fa1-649c7a58fee9.png">

mmtr commented:
> I did notice that on the JN site, upgrading the plan resulted in a new field, `author_wpcom`, for existing media items. New VideoPress media items, however, don't have this field; I'm not sure if that's related, or a separate bug (or not?) altogether. 

Nice catch! It seems it is not related to this. The `author_wpcom` field was introduced in D17007-code and it appears when the site is connected. 

In fact, I can see it in all the media items, even before upgrading the plan. Are you sure this field were not present on your test before upgrading the plan?

I can reproduce however the lack of this field on new VideoPress media items. Apparently this is expected, because the `author_wpcom` field is included only if the author is mapped to a WP.com user, but the VideoPress items are created by an automated process, so their author is actually the user with ID 0 which is not mapped to any WP.com user.

> Also, two media items were created by just one upload, one with JP data and one without – this doesn't happen on dotcom, however. Oh! One is a jpeg, so maybe a poster image of some sort? But still, the same video file did not create this second media item on dotcom.

Yeah, this is expected and caused by a divergence between WP.com and Jetpack. The VideoPress module in Jetpack automatically creates a new media item for the poster image. In WP.com however, no poster image is created.

gwwar commented:
Would it be possible to add unit tests to cover these new cases?

mmtr committed:
Move plugin to wpcom-fields

mmtr committed:
Add tests

mmtr commented:
> > Should this be in `wpcom-fields` instead of `wpcom-endpoints`? That seems more appropriate, I think (although I couldn't really find any documentation on the difference, this just seems to be doing what `_inc/lib/core-api/wpcom-fields/post-fields-publicize-connections.php` is doing too).
> 
> I don't think so. We already have `sites-posts-featured-media-url.php` in `wpcom-endpoints` which also includes a new field on the API REST responses.

Actually, you were right @kwight.  After taking a deeper look,`wpcom-fields` is the place for adding logic that needs to register new fields using `register_rest_field` so I moved the code there.

mmtr commented:
> Would it be possible to add unit tests to cover these new cases?

I added them, but they're failing right now and I'm not sure how to fix them. The code relies on the existing `video_get_info_by_blogpostid` and `videopress_get_video_details` functions, so I think it is ok to mock them and check on the test we're actually using what these functions return for the new `jetpack_videopress` field.

The problem is that these functions are global and I don't now how to mock them. I tried to use namespaces as suggested [here](https://www.codepunker.com/blog/overwrite-built-in-php-functions-using-namespaces) but it doesn't work.

If we don't mock these functions and since `videopress_get_video_details` retrieves the data externally from the public API, we would really need to upload a video to VideoPress and wait until it is transcoded, which I think is less than ideal.

gwwar commented:
@mmtr if you get stuck with testing I'll allow a unit or full E2E test as a follow up so as to not block this PR.

mmtr committed:
Refactor class methods so they can be mocked in unit tests

mmtr committed:
Fix unit tests

mmtr commented:
thanks @gwwar! I managed to add some working tests, so this should be ready now.

gwwar reviewed (dismissed):
Overall this tests well for me @mmtr ! Thanks for adding some helpful comments around the permission check functions!

Non-blocking, but I did notice this property is added to non-video files, for example images. Would we like to keep this consistent for the media response or have this as an additional property for videos?

<img width="1095" alt="screen shot 2019-01-30 at 2 34 06 pm" src="https://user-images.githubusercontent.com/1270189/52017566-8e281380-249c-11e9-8ab8-afb75b6617f5.png">

<img width="313" alt="screen shot 2019-01-30 at 2 37 30 pm" src="https://user-images.githubusercontent.com/1270189/52017635-caf40a80-249c-11e9-93b3-9946daa3a6e0.png">

mmtr commented:
> Non-blocking, but I did notice this property is added to non-video files, for example images. Would we like to keep this consistent for the media response or have this as an additional property for videos?

@gwwar yeah, ideally I'd like to add the field only to videos, but `WPCOM_REST_API_V2_Field_Controller` uses `register_rest_field` for defining the object the field is being registered to. 

I believe we don't have any specific object type in WP for the videos, and `attachment` is the only one we can use. The alternative would be to build a new endpoint specific for videos that will mimics several parts from the attachments endpoints (i.e. check permissions, check if media exists, etc) which is why I decided to extend the existing attachments endpoint (less code to maintain).

jeherve reviewed (dismissed):
This works well overall. I only had one minor comment, and a more general question about your approach.

With your implementation, I may use VideoPress and a Jetpack plan for a year, and then choose not to renew my plan. VideoPress will thus be deactivated once the plan is not available on the site anymore, but the videos uploaded to the site during the year remain on the site and are still embedded in posts and pages. Those videos should keep working.

It consequently makes me wonder if instead of having meta data that is fetched from WordPress.com and thus only available when the module is active, we should have that meta data saved to the attachment locally, on the site (with something like `register_meta`, triggered when the video is actually added to the media library in the first place). This way the data would remain available as long as the attachment remains, even when VideoPress is not in use anymore.

That's just a question though, I may be missing something obvious here :)

mmtr committed:
Include VideoPress data field always in API responses.

This way, the data would remain available as long as the attachment
remains, even when VideoPress is not in use anymore.

mmtr commented:
@jeherve nice insight! In fact, such metadata is already stored, so this can be easily solved by always loading this plugin, without checking if the VideoPress module is activated or not. I changed that on af207c4.

mmtr commented on _inc/lib/core-api/wpcom-fields/attachment-fields-videopress.php:
```diff
@@ -0,0 +1,114 @@
+}
+
+if (
+	( method_exists( 'Jetpack', 'is_module_active' ) && Jetpack::is_module_active( 'videopress' ) ) ||
```
Added in af207c4

dmsnell commented:
Just noting here that `WPCOM_REST_API_V2_Field_Controller` is a thin wrapper around [`register_rest_field()`](https://developer.wordpress.org/reference/functions/register_rest_field/) from core and there are some familiar oddities about it:

 - outside of this second instance it's only used in one place as far as I can tell
 - there are demands on implementing the interface that both aren't relevant to the work we're doing (permission checks, for example) _and_ which the Core function doesn't demand

mmtr commented:
@dmsnell yeah, I don't like neither to override so many methods we really don't need. Once we merge this PR, I'll open a follow-up PR refactoring the code so we use `register_rest_field` directly instead of of `WPCOM_REST_API_V2_Field_Controller`.

mmtr commented:
@jeherve just a reminder that I addressed your last feedback and this should be ready for a final review 🙂

mdawaffe reviewed (commented):
I second @gwwar's earlier [comment](https://github.com/Automattic/jetpack/pull/11162/#pullrequestreview-198344669). Can we not output this field when it doesn't apply?

Here's a proposal. I'm not sure if the `is_video()` check is good enough and/or if it should really be an `is_videopress_attachment()` check that does something more intelligent. Perhaps we should just check if the value is `null` and remove the field if so :)

```diff
diff --git a/_inc/lib/core-api/wpcom-fields/attachment-fields-videopress.php b/_inc/lib/core-api/wpcom-fields/attachment-fields-videopress.php
index 1b1b7a0bf..4d0a1af29 100644
--- a/_inc/lib/core-api/wpcom-fields/attachment-fields-videopress.php
+++ b/_inc/lib/core-api/wpcom-fields/attachment-fields-videopress.php
@@ -15,6 +15,12 @@ class WPCOM_REST_API_V2_Attachment_VideoPress_Field extends WPCOM_REST_API_V2_Fi
 	protected $object_type = 'attachment';
 	protected $field_name  = 'jetpack_videopress';
 
+	public function register_fields() {
+		parent::register_fields();
+
+		add_filter( 'rest_prepare_attachment', array( $this, 'remove_field_for_non_videos' ), 10, 2 );
+	}
+
 	/**
 	 * Defines data structure and what elements are visible in which contexts
 	 */
@@ -53,6 +59,18 @@ class WPCOM_REST_API_V2_Attachment_VideoPress_Field extends WPCOM_REST_API_V2_Fi
 		return $videopress_data;
 	}
 
+	public function is_video( $attachment ) {
+		return 0 === strpos( $attachment->post_mime_type, 'video/' );
+	}
+
+	public function remove_field_for_non_videos( $response, $attachment ) {
+		if ( ! $this->is_video( $attachment ) ) {
+			unset( $response->data[$this->field_name] );
+		}
+
+		return $response;
+	}
+
 	/**
 	 * Get VideoPress data for a given attachment.
 	 *
```

mdawaffe commented:
> Just noting here that `WPCOM_REST_API_V2_Field_Controller` is a thin wrapper around [`register_rest_field()`](https://developer.wordpress.org/reference/functions/register_rest_field/) from core and there are some familiar oddities about it:
> 
> * outside of this second instance it's only used in one place as far as I can tell
> * there are demands on implementing the interface that both aren't relevant to the work we're doing (permission checks, for example) _and_ which the Core function doesn't demand

`WPCOM_REST_API_V2_Field_Controller` provides some handy things like schema enforcement for non-scalar objects and context checks for the same that core doesn't handle correctly. I recommend sticking with `WPCOM_REST_API_V2_Field_Controller` in this case because of the field's complex value (an object).

The permissions stuff is more useful for things that are not read-only :) but, yes, it does feel overboard for read-only stuff like this. Thanks for pointing it out, @dmsnell, @mmtr. Perhaps we can iterate on it to make it better. Maybe by defining `'readonly' => true` in the schema, we could skip those checks? Probably best as a future improvement in another PR :)

mmtr committed:
Remove the field for non videos

mmtr committed:
Make sure the field is always a string

mmtr committed:
Remove default value

mmtr committed:
Add description to schema

mmtr commented:
Thanks for the review @mdawaffe! I added your proposal of removing the field for non video. Thanks for the suggestion on how to do it!

mmtr commented on _inc/lib/core-api/wpcom-fields/attachment-fields-videopress.php:
```diff
@@ -0,0 +1,114 @@
+	/**
+	 * Defines data structure and what elements are visible in which contexts
+	 */
+	public function get_schema() {
```
I realized that we only need the guid, so I simplified the schema and the field returned to be only a string.

mmtr commented on _inc/lib/core-api/wpcom-fields/attachment-fields-videopress.php:
```diff
@@ -0,0 +1,114 @@
+		return array(
+			'$schema'    => 'http://json-schema.org/draft-04/schema#',
+			'title'      => $this->field_name,
+			'type'       => 'object|null',
```
Changed to `'string'` since we're returning now just the guid.

jeherve reviewed (changes requested):
It seems to work well for me. I only have one minor i18n comment.

mmtr committed:
Add textdomain to translation

mmtr commented on _inc/lib/core-api/wpcom-fields/attachment-fields-videopress.php:
```diff
@@ -0,0 +1,171 @@
+			'type'        => 'string',
+			'context'     => array( 'view', 'edit' ),
+			'readonly'    => true,
+			'description' => __( 'Unique VideoPress ID' ),
```
Yes, it should be. We already translate other schema descriptions ([#](https://github.com/Automattic/jetpack/blob/master/_inc/lib/core-api/wpcom-endpoints/gutenberg-available-extensions.php#L48), [#](https://github.com/Automattic/jetpack/blob/ca4562452e700db84da3337dbdbeb0ddcdde36da/_inc/lib/core-api/wpcom-fields/post-fields-publicize-connections.php#L73)), but I forgot to include the textdomain, thanks for pointing that out. Added in 043e229

mmtr committed:
Add textdomain to translation checked in unit tests

kwight commented on _inc/lib/core-api/wpcom-fields/attachment-fields-videopress.php:
```diff
@@ -0,0 +1,171 @@
+	 * @return false|int
+	 */
+	public function is_video( $attachment ) {
+		return preg_match( '@^video/@', $attachment->post_mime_type );
```
Curious why you switched to `preg_match` from the initial suggestion of `strpos`? The latter is considered [[ http://php.net/manual/en/function.preg-match.php#refsect1-function.preg-match-notes | more performant and preferable ]] when checking for a string within a string. I'm also not familiar with the `@`, nor could I quickly find a reference; what is its purpose?

mdawaffe replied:
The `@` is just the delimiter. In PCRE (what PHP uses for regex's, hence: **p**reg...), you can use any most any "special" character: http://php.net/manual/en/regexp.reference.delimiters.php

We tend to use `/` for consistency, but many characters will do. Here, it's convenient to use something that is not `/` since, otherwise, we'd have to escape the `/` in the pattern itself.

If the switch to `preg_match()` was made for readability, we can also use `wp_startswith( $haystack, $needle )` in WP.com and Jetpack, which may meet the readability requirement while still using the more performant `strpos()`.

dmsnell replied:
concerning the `@` PHP is pretty permissive in which [delimiters](http://php.net/manual/en/regexp.reference.delimiters.php) it uses for RegExp patterns - well, this probably goes back to the PERL roots

you can basically use almost anything there as long as it balances on the end.

also, the transformed version of this match…

```php
return 0 === strpos( $attachment->post_mime_type, 'video/' );
```

mmtr replied:
> Curious why you switched to `preg_match` from the initial suggestion of `strpos`?

Just for readability and because I noted we are already using that in other places ([#](https://github.com/Automattic/jetpack/blob/7b7d869dbaa6131e8782431c0f78638cc990ea98/modules/videopress/class.jetpack-videopress.php#L302), [#](https://github.com/Automattic/jetpack/blob/7b7d869dbaa6131e8782431c0f78638cc990ea98/json-endpoints/class.wpcom-json-api-upload-media-v1-1-endpoint.php#L167)).

Didn't know about the existence of `wp_startswith` (thanks @mdawaffe!) so I'll commit now a change using it :).

mmtr replied:
> In the current implementation, it's added for all videos. Is that the desired behavior?

That's right. 

From an API consumer perspective, I think it is more natural to have the field in all videos, since all of them can potentially be VideoPress videos. Therefore, the value of this field will indicate if it is a VideoPress video or not.

This also makes the implementation easier because we don't need to check if the field is present or not.

jeherve reviewed (dismissed):
This works well in my tests. 👍

dmsnell commented on _inc/lib/core-api/wpcom-fields/attachment-fields-videopress.php:
```diff
@@ -0,0 +1,171 @@
+	 *
+	 * @return string
+	 */
+	public function get( $object, $request ) {
```
we have some type hinting that has been available since PHP 5.0.0 and PHP 5.1.0 - if we're going to use `$object['id']` we might consider typing it to remove confusion

```php
public function get( array $object, WP_REST_Request $request ) {
	…
}
```

this isn't a blocker in any way

mmtr replied:
Thanks for the suggestion! Since we already have the docs in the function, I decided to rename the variable and change the `@param` type to array, so it should be clear now.

dmsnell commented on _inc/lib/core-api/wpcom-fields/attachment-fields-videopress.php:
```diff
@@ -0,0 +1,171 @@
+	/**
+	 * Gets the VideoPress GUID for a given attachment.
+	 *
+	 * This is pulled out into a separate method to support unit test mocking.
```
you are so amazing - thanks for clearly indicating why this thin wrapper exists!

dmsnell commented on _inc/lib/core-api/wpcom-fields/attachment-fields-videopress.php:
```diff
@@ -0,0 +1,171 @@
+		}
+
+		return $response;
+	}
```
is there no way to conditionally add it vs. always adding it and conditionally removing it?

I'm guessing this isn't possible given the API here?

mdawaffe replied:
I couldn't figure out a way :( (with this abstraction or directly with core's `register_rest_field()`). The filters just aren't quite there.

mmtr committed:
Use wp_startswith instead of preg_match

mmtr committed:
Change variable name and docs to make clear is an array

mmtr commented:
I rebased the PR in order to fix some failing tests and that caused the dismissal of @jeherve's review (it was not intentional!). 

Given that, I took the opportunity to commit a couple of minor changes in order to address the feedback from https://github.com/Automattic/jetpack/pull/11162#discussion_r254922051 and https://github.com/Automattic/jetpack/pull/11162#discussion_r254902723.

gwwar reviewed (approved):
Thanks @mmtr I followed testing instructions and the API responses appeared as expected! I did notice some locally uploaded videos (not stored on videopress) return a response of `jetpack_videopress_guid: ""`

For clients is an empty string here okay? Or should we be more explicit and return `null` or remove the prop? I'll note it as non-blocking, but wanted to point out the possibility of a client checking for a falsey value in an incorrect way.

mdawaffe commented:
Not sure which is better: returning `""` or removing the field (from https://github.com/Automattic/jetpack/pull/11162#discussion_r255021893, it sounds like the empty string is better). I suggest not returning `null` so that we always return the same data type: a string.

mmtr commented:
> For clients is an empty string here okay?

I'd say so and it was my reasoning on https://github.com/Automattic/jetpack/pull/11162#discussion_r255021893. 

Also, I agree with @mdawaffe that we should return always the same data type.

jeherve reviewed (approved):
This is looking good on my end. 👍 	🚢
